### DISCUSSION TITLE
Relatable attribute methods not triggering

### DISCUSSION THREAD

**Discussion by feeh27** [laravel/nova-issues#5610] [Q&A] [answered]
- Laravel Version: 10.13.1
- Nova Version: 4.25.1
- PHP Version: 8.2.6
- Database Driver & Version: SQLite
- Operating System and Version: Windows 11 Pro 64 bits
- Browser type and version: Google Chrome - Version 114.0.5735.110
- Reproduction Repository: https://github.com/feeh27/nova-issues-5608

### Description:

Relatable attribute methods are not working, I need to filter the belongsTo options, because I have a field called "parent" with self-reference and when I edit a resource, I don't want it to appear as an option.

Official doc: https://nova.laravel.com/docs/4.0/resources/authorization.html#relatable-filtering

I have the following resource classes and none of the relatableParent or relatableParents methods are called:

Resource:
```php
// Fields

public function fields(NovaRequest $request)
{
    return [
        ID::make()->sortable(),

        BelongsTo::make('Parent', 'parent', static::class)
            ->sortable()
            ->nullable()
            ->withoutTrashed(),

        Text::make('Name')
            ->sortable()
            ->rules('required'),

        Select::make('Condition')
            ->sortable()
            ->rules(['required', 'in:and,or'])
            ->displayUsingLabels()
            ->options(fn () => [
                ['value' => 'and', 'label' => 'And'],
                ['value' => 'or', 'label' => 'Or'],
            ]),

        Select::make('Operator')
            ->sortable()
            ->rules(['required', 'in:eq,ne,in'])
            ->displayUsingLabels()
            ->options(fn () => [
                ['value' => 'eq', 'label' => 'Equals'],
                ['value' => 'ne', 'label' => 'Not Equals'],
                ['value' => 'in', 'label' => 'In'],
            ]),

        Text::make('Field')
            ->sortable()
            ->rules('required'),

        Text::make('Value')
            ->sortable()
            ->rules('required'),
    ];
}

// Relatable attribute methods

public static function relatableParent(NovaRequest $request, $query)
{
    $resourceId = $request->route('resourceId');

    return $query->whereNotIn('id', [$resourceId]);
}

public static function relatableParents(NovaRequest $request, $query)
{
    $resourceId = $request->route('resourceId');

    return $query->whereNotIn('id', [$resourceId]);
}
```

Model:
```php
<?php

declare(strict_types=1);

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\SoftDeletes;

class QueryBuilder extends Model
{
    use SoftDeletes;

    protected $table = 'query_builder';

    protected $fillable = [
        'parent_id',
        'name',
        'condition',
        'operator',
        'value',
    ];

    public function parent(): BelongsTo
    {
        return $this->belongsTo(
            static::class,
            'parent_id',
            'id',
        );
    }
}
```

Even with the methods configured according to the documentation, the method is not called and the "Filter 3" record is still available for selection, as shown in the image below:

![image](https://github.com/laravel/nova-issues/assets/13547485/db1a9613-a114-4feb-b07b-a15e81d3d16a)

### Detailed steps to reproduce the issue on a fresh Nova installation:

- Login
- Go to Query Builder page
- Create one or more records
- Select a record and go to its edit page
- Click in Parent field and all records will be present

**Comment by crynobone** [comment]
`resourceId` here is from the query string and not route.

**Reply by feeh27** [reply]
From official doc:
![image](https://github.com/laravel/nova-issues/assets/13547485/49efe515-3a13-416d-9003-fff4a5913398)

Link: https://nova.laravel.com/docs/4.0/resources/authorization.html#relatable-filtering

**Reply by feeh27** [reply]
I tested it on the relatableQuery method and the resourceId is returned correctly

**Reply by crynobone** [reply]
Will fix the documentation, sorry for the inconvenience: https://github.com/laravel/nova-docs/pull/528

**Comment by feeh27** [comment]
The problem is still the same, relatable attribute methods aren't called.

**Reply by crynobone** [reply]
Just added dusk tests to verify current usage should work fine:

https://github.com/laravel/nova-dusk-suite/blob/c49b62edbdccc04436f4012b9a97c6ffb7029317/tests/Browser/RelationshipAuthorizationTest.php#L191-L208

https://github.com/laravel/nova-dusk-suite/blob/c49b62edbdccc04436f4012b9a97c6ffb7029317/app/Nova/Employee.php#L102-L114

If you do have an issue we need to have a full reproducing repository to show the issue as your issue might be due to 3rd party package overriding the default behavior causing it not to be used (similar report in the past).

**Reply by feeh27** [reply]
I don't understand why my issue was converted into a discussion, since it is indeed a bug, in the original issue there is a link to a repository using a recent version of Nova, without third party packages and in which the reported behavior happens.

https://github.com/feeh27/nova-issues-5608

**Reply by crynobone** [reply, marked as answer]
It should have been `relatableQueryBuilders()`

**Reply by feeh27** [reply]
Have you read the official documentation? (I gave her link here) They are completely different things, the error persists and it is a bug or Laravel Nova is making a big mistake in the documentation of a feature.

**Reply by feeh27** [reply]
Any updates? @crynobone

**Reply by crynobone** [reply]
See https://github.com/laravel/nova-docs/pull/529

**Reply by feeh27** [reply]
I don't need the documentation tweak, my problem is with Dynamic Relatable Methods and not with the relatableQuery.

**Reply by crynobone** [reply]
Feel free to ignore the answer, have a good day.

**Reply by feeh27** [reply]
Sorry for the misunderstanding, the answer below solves the problem.

I believed you were confusing it with `relatableQuery`, but it was actually my confusion.

> It should have been `relatableQueryBuilders()`

**Reply by feeh27** [reply]
I'll leave here @crynobone's comment that is in the documentation correction PR so that other developers with the same problem can find the solution:

> Prevent confusion when `relatableTags` was assumed to be `relatable{Resource}s` but instead uses `relatable{Model}s`.
> 
> Using `relatable{Model}s` is inline with all dynamic methods available in this page and we don't use anything referencing `{Resource}` anywhere in Nova.
